src/core/ddsc/tests/psmx.c

```diff
@@ -1264,6 +1391,164 @@ CU_Test (ddsc_psmx, basic)
+/// - Create some entities
+/// - Check if shared memory is enabled.
+/// - Expectation: Shared memory is enabled iff the psmx interface supports it (use queried value).
+/// - Delete the domain
+/// - Check the function call counts of the dummy psmx.
+/// - Expectation: The counts match expectations. In particular, create counts must match their delete counterpart.
+/// 
+CU_Test(ddsc_psmxif, shared_memory)
+{
+  const char* CDDS_PSMX_NAME = NULL;
+  ddsrt_getenv("CDDS_PSMX_NAME", &CDDS_PSMX_NAME);
+  ddsrt_setenv("CDDS_PSMX_NAME", "dummy"); // Make `create_participant()` use the dummy psmx.
+  dummy_mockstats_t dmock, dmock_expected;
+  char strbuf1[512];
+  char strbuf2[512];
+  size_t strbuf_size = sizeof(strbuf1);
+  {
+    // Check that the data types I'm planning to use are actually suitable for use with shared memory.
+    dds_data_type_properties_t props;
+    props = dds_stream_data_types(SC_Model_desc.m_ops);
+    CU_ASSERT_FATAL((props & DDS_DATA_TYPE_IS_MEMCPY_SAFE) == DDS_DATA_TYPE_IS_MEMCPY_SAFE);
+    props = dds_stream_data_types(PsmxType1_desc.m_ops);
+    CU_ASSERT_FATAL((props & DDS_DATA_TYPE_IS_MEMCPY_SAFE) == DDS_DATA_TYPE_IS_MEMCPY_SAFE);
+  }
+
+  const dds_domainid_t domainId = 3;
+  bool supports_shared_memory = false; // Overwritten by `plugin's supported_features()`.
+  bool specify_locator[] = {true, false};
+  uint8_t locator_in[16];
+  memset(locator_in, 0x0, sizeof(locator_in)); // Avoid warning 'uninitialized value'.
+  ((uint64_t*)locator_in)[0] = (uint64_t)0x4a4d203df6996395;
+  ((uint64_t*)locator_in)[1] = (uint64_t)0xe1412fbecc2de4b6;
+
+  int N = sizeof(specify_locator) / sizeof(specify_locator[0]);
+  for (int i = 0; i < N; ++i)
+  {
+    const uint8_t* locator = specify_locator[i] ? locator_in : NULL;
+    dds_entity_t participant = create_participant(domainId, locator);
+    CU_ASSERT_FATAL(participant > 0);
+    dds_entity_t domain = dds_get_parent(participant);
+    CU_ASSERT_FATAL(domain > 0);
+    if ( specify_locator[i] ) {
+      // Check that I get the same locator that I provided with the config.
+      ddsi_locator_t locator_out;
+      CU_ASSERT_FATAL(domain_get_psmx_locator(domain, &locator_out));
+      CU_ASSERT_FATAL(memcmp(locator_in, locator_out.address, sizeof(locator_out.address)) == 0);
+    }
+
+    {
+      // Query whether shared memory is supported.
+      dds_domain* dom = NULL;
+      CU_ASSERT_FATAL(domain_pin(domain, &dom));
+      CU_ASSERT_FATAL(dom->psmx_instances.length == 1); // There shall be exactly one psmx instance.
+      struct dummy_psmx* dpsmx = (struct dummy_psmx*)dom->psmx_instances.instances[0];
+      dpsmx->mockstats_get_ownership(&dmock);
+
+      // The psmx must have an instance_name that is not an empty string.
+      CU_ASSERT_FATAL(dpsmx->c.instance_name != NULL && strcmp(dpsmx->c.instance_name, "") != 0);
+      supports_shared_memory = ((dpsmx->c.ops.supported_features(&dpsmx->c) & DDS_PSMX_FEATURE_SHARED_MEMORY) == DDS_PSMX_FEATURE_SHARED_MEMORY);
```

Comment:
I would do this by calling `dds_is_shared_memory_available` on a reader/writer. What you want to know is that that function calls the plugin's `supported_features` and that its return value corresponds to what the plugin replied.

This merely checks the dummy plugin implementation, not the core.